packages/optimistic-oracle/src/keeper.js

```diff
@@ -0,0 +1,250 @@
+    this.fromWei = this.web3.utils.fromWei;
+    this.utf8ToHex = this.web3.utils.utf8ToHex;
+    this.hexToUtf8 = this.web3.utils.hexToUtf8;
+
+    // Default config settings. Liquidator deployer can override these settings by passing in new
+    // values via the `ooKeeperConfig` input object. The `isValid` property is a function that should be called
+    // before resetting any config settings. `isValid` must return a Boolean.
+    const defaultConfig = {
+      txnGasLimit: {
+        // `txnGasLimit`: Gas limit to set for sending on-chain transactions.
+        value: 9000000, // Can see recent averages here: https://etherscan.io/chart/gaslimit
+        isValid: x => {
+          return x >= 6000000 && x < 15000000;
+        }
+      }
+    };
+
+    // Validate and set config settings to class state.
+    const configWithDefaults = createObjectFromDefaultProps(ooKeeperConfig, defaultConfig);
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
+    await Promise.all([this.optimisticOracleClient.update(), this.gasEstimator.update()]);
+
+    // Increase allowances for all relevant collateral currencies.
+    // TODO: Consider whether this should happen in a separate function (i.e. `setAllowances`) or within
+    // the `sendProposals()` method.
+    await this._setAllowances();
+  }
+
+  // Sets allowances for all collateral currencies used in unproposed price requests
+  async _setAllowances() {
+    // TODO: Should we do this also for currencies from undisputedProposals?
+    for (let priceRequest of this.optimisticOracleClient.getUnproposedPriceRequests()) {
+      // The OO requires approval to transfer the proposed price request's collateral currency in order to post a bond.
+      // We'll set this once to the max value and top up whenever the bot's allowance drops below MAX_INT / 2.
+      const collateralToken = new this.web3.eth.Contract(getAbi("ExpandedERC20"), priceRequest.currency);
+      const [currentCollateralAllowance] = await Promise.all([
+        collateralToken.methods.allowance(this.account, this.ooContract.options.address).call()
+      ]);
+      if (this.toBN(currentCollateralAllowance).lt(this.toBN(MAX_UINT_VAL).div(this.toBN("2")))) {
+        const collateralApprovalTx = await collateralToken.methods
+          .approve(this.ooContract.options.address, MAX_UINT_VAL)
+          .send({
+            from: this.account,
+            gasPrice: this.gasEstimator.getCurrentFastPrice()
+          });
+        this.logger.info({
+          at: "OptimisticOracle#Keeper",
+          message: "Approved OO to transfer unlimited collateral tokens 💰",
+          currency: collateralToken.options.address,
+          collateralApprovalTx: collateralApprovalTx.transactionHash
+        });
+      }
+    }
+  }
+
+  // Create the pricefeed for a specific identifier and save it to the state, or
+  // return the saved pricefeed if already constructed.
+  async _createOrGetCachedPriceFeed(identifier) {
```

Comment:
nit: do you want to put these internal methods at the bottom of the class?

Reply:
Will do